packages/queued-request-controller/src/QueuedRequestController.ts

```diff
@@ -263,30 +282,64 @@ export class QueuedRequestController extends BaseController<
    */
   async #processNextBatch() {
     const firstRequest = this.#requestQueue.shift() as QueuedRequest;
-    this.#originOfCurrentBatch = firstRequest.origin;
-    this.#networkClientIdOfCurrentBatch = firstRequest.networkClientId;
-    const batch = [firstRequest.processRequest];
+    this.#originOfCurrentBatch = firstRequest.request.origin;
+    this.#networkClientIdOfCurrentBatch = firstRequest.request.networkClientId;
+    const batch = [firstRequest];
+
+    let networkSwitchError: unknown;
+    try {
+      // If globally selected network is different from origin selected network,
+      // switch network before processing batch
+      await this.#switchNetworkIfNecessary(
+        firstRequest.request.networkClientId,
+      );
+    } catch (error: unknown) {
+      networkSwitchError = error;
+    }
+
+    // If the first request might switch the network, process the request by
+    // itself. If the request does change the network, clear the queue for the
+    // origin since it any remaining requests are now invalidated
+    if (this.#canRequestSwitchNetworkWithoutApproval(firstRequest.request)) {
+      // This hack prevents the next batch from being processed
+      // after this request returns. This is necessary because
+      // we may need to flush the queue before the next set of requests
+      // are batched and processed, which we cannot do without blocking
+      // the queue from continuing by artificially increasing the processing
+      // request count
+      this.#processingRequestCount += 1;
+      try {
+        firstRequest.processRequest(networkSwitchError);
+        this.#updateQueuedRequestCount();
+        await firstRequest.requestHasBeenProcessed;
+      } finally {
+        this.#processingRequestCount -= 1;
+      }
+      const { selectedNetworkClientId } = this.messagingSystem.call(
+        'NetworkController:getState',
+      );
+      if (this.#networkClientIdOfCurrentBatch !== selectedNetworkClientId) {
+        this.#flushQueueForOrigin(this.#originOfCurrentBatch);
+      }
+      this.#processNextBatchIfReady();
```

Comment:
Nit: here's a suggestion for a commend to explain why this is called

```suggestion
      // Re-trigger processing of next batch because the `this.#processingRequestCount` guard above
      // prevents it from being triggered when it typically would, after the request resolves.
      this.#processNextBatchIfReady();
```